### PR TITLE
A simple plugin system

### PR DESCRIPTION
**Pull request by roveo**
Right now, if I need to add my own custom stream nodes, I have to do this:

```py
import mypackage.streamz_extras  # noqa: F401
```

It would be nice to have a way to distribute additional functionality as separate packages that can just be installed via pip. This can be done with entry_points, similar to the way it works in airflow. This is a super bare-bones implementation of this mechanism. Check out https://github.com/roveo/streamz_example_plugin for an example of a plugin.

Problems:
- this should be tested, but I have no idea how, short of shipping the code for example plugin with tests
- plugged-in classes should be checked for validity in some way. I can add a simple check e.g. `isinstance(plugin, Stream)`, but there is probably something else I haven't thought of

### CODE REVIEW

**Comment by martindurant** [comment]
Thank you. I totally agree that using entry_points is a great alternative or addition to the registration system.
@CJ-Wright @jsmaupin @chinmaychandak - anyone have time and interest to give this a once-over?

**Comment by jsmaupin** [comment]
I think this is a great idea.  I was hoping that we could figure out how do this in the future for sources and sinks where you could connect Streamz to external services using packages install-able via pip or conda because, these connections often have a lot of dependencies that you don't necessarily want in your core package.  I hadn't considered it for functionality that operates in the middle of the pipeline. I think this is the step that gets us in that direction.

**Comment by CJ-Wright** [comment]
I think this is great! How should we document all these plugins so people can discover them?

**Comment by martindurant** [comment]
If we go with entry_points as a way of declaring plugins, then we *do not* document them, at least not exhaustively. Intake faces this: we have a page of known plugins, which people can submit PRs for changing, but there's nothing to stop new plugins appearing which we don't know about. It would be possible to have a configuration that only loads a select number of plugins, or instead to not load the plugins except upon explicit user invocation.

**Comment by chinmaychandak** [comment]
I think is really nice! 

I think for validating the plugins, we would definitely need `isinstance(plugin, Stream)` (similar to Airflow which tests against one of its subclasses called `PluginManager` which I don't think is necessary at this point).  

Also, for testing an example plugin, having a separate repository/page called `streamz-plugins` could be an option? I think that's what Airflow does, too. The `from_iterable` node you wrote can be a good plugin for testing. 

As for tests for future plugins, I think if we consider adding a plugin to `setup.py` in streamz, then we would need to verify the plugin behavior separately beforehand; if, however, we allow the plugin to be loaded lazily (only when the user explicitly invokes them, either through CLI or otherwise), then I think we don't need to worry about having tests in streamz itself (the plugin repo can have those tests, I mean). I'm just speaking my mind here, I may be missing something!

**Comment by roveo** [comment]
> Also, for testing an example plugin, having a separate repository/page called streamz-plugins could be an option?
I think this is the way to go. Then I'll add tests, so that they are not bound to the repo in my account.

> As for tests for future plugins, I think if we consider adding a plugin to setup.py in streamz, then we would need to verify the plugin behavior separately beforehand
So, plugins in streamz-plugins repo that are tested and "verified" can be provided via `extras_require` in setup.py, but nothing is stopping anyone from creating their own plugin package and installing it with pip.

**Comment by chinmaychandak** [comment]
This seems reasonable to me, although I would defer to Martin and CJ to make the final call whether or not this is the best way to proceed.

**Comment by chinmaychandak** [comment]
I see Travis builds timing out. Is this due to it being overloaded? Should we try something like `install: travis_wait 30 mvn install` temporarily?

**Comment by martindurant** [comment]
We need to transfer our tests away from travis...

**Comment by martindurant** [comment]
Yes, we should do whatever it takes to get travis to pass for now...

I am only now looking at the actual code. I wonder, is there any appetite for making this system lazy? As it is, streamz will import all packages that claim to have relevant plugins - but the node classes still need to be annotated with decorators as before.

I could imagine entry points spelled like:
```
streamz.plugins: [
  "transform=streamz_addon.module.TransformNode"
]
```
which adds "transform" to the dir() of Stream, but only imports and registers when the `.transform` is first used (i.e., via a `__getattr__`).

Furthermore, we may want to be more specific with our entrypoint group naming: 
- streamz.node
- streamz.source
- streamz.sink
- ...

What do you think?

btw: does this require the package `entrypoints` as a dependency?

**Comment by CJ-Wright** [comment]
I like having them lazy, since that allows us to not need the plugins to be installed until we actually need them

**Comment by CJ-Wright** [comment]
@martindurant would it be worthwhile to setup the CI using conda-forge's CI provisioning system? That way if we need to move to something we just rerender?

**Comment by martindurant** [comment]
Sorry @CJ-Wright , you're the expert there, I really don't know what's involved.

**Comment by CJ-Wright** [comment]
I'll put up a PR if we're interested, but the rough outline is here: https://conda-forge.org/docs/user/ci-skeleton.html

**Comment by martindurant** [comment]
Does it work for github actions CI?

**Comment by CJ-Wright** [comment]
Not yet, since CF doesn't use GH actions as a CI (yet). The hope of this approach would be that we don't need to care about what particular CI we are using, since we could rerender and the CIs could move.

We could add GH actions to CF and then it would support that CI.

**Comment by martindurant** [comment]
In that case I am hesitant, since the dask repos are all going to github. I don't have a circle or azure account myself - is it all free like Travis used to be?

**Comment by CJ-Wright** [comment]
Sorry, can we move the CI conversation to an issue? I don't want to derail things here. https://github.com/python-streamz/streamz/issues/382

**Comment by roveo** [comment]
> As it is, streamz will import all packages that claim to have relevant plugins - but the node classes still need to be annotated with decorators as before.

Do you mean that with lazy loading they won't have to be annotated?

> adds "transform" to the dir() of Stream, but only imports and registers when the .transform is first used (i.e., via a __getattr__)

This can be done with a @property that looks it up in a class attribute dict (e.g. Stream._plugins) and loads the entrypoint if it's not already loaded.

> Furthermore, we may want to be more specific with our entrypoint group naming

Do these types of nodes require different treatment when being loaded?

> btw: does this require the package entrypoints as a dependency?

I don't think so, this is built into setuptools. `entrypoints` looks like a different package (with the last 2 releases in 2017 and 2019).

**Comment by martindurant** [comment]
> Do you mean that with lazy loading they won't have to be annotated?

Indeed, this would be an alternative to the annotation; although doing both should be allowed for backward compatibility. 

> This can be done with a @property that looks it up in a class attribute dict (e.g. Stream._plugins) and loads the entrypoint if it's not already loaded.

Something like that, yes. There are many ways to achieve it. You would call register_api on the class once it's imported, so it gets added to the namespace and you don't need to import it again.

> Do these types of nodes require different treatment when being loaded?

*maybe*. This gives us more flexibility for the future. We could argue that, for the time being, all possible plugins as "nodes" (i.e., they appear as attributes of Stream). My motivation here is, that in Intake we started off having things called "plugins", but then made other parts of the library pluggable and had to, painfully, rename the original plugins to "drivers".

> I don't think so, this is built into setuptools.

Better check that setuptools has this functionality for all python versions of interest. Also, setuptools had better be included in the deps (yes, I know everyone already has it).

**Comment by roveo** [comment]
> You would call register_api on the class once it's imported

Then register_api would have to allow us to specify which attribute the class will be registered at. Right now it's `setattr(cls, func.__name__, modifier(wrapped))`, so it's always the class name.

> Better check that setuptools has this functionality for all python versions of interest. Also, setuptools had better be included in the deps (yes, I know everyone already has it).

OK, will do.

**Comment by martindurant** [comment]
Correct. I imagine we'll have redundancy for now "name=package.module.name", but there might be a case for when we want the class name and method name to be different, which is not yet possible.

Actually, there is a difference between normal nodes and sources: the latter are registered as `staticmethod`.

**Comment by chinmaychandak** [comment]
@roveo, could you try something like `install: travis_wait 30 mvn install` in `travis.yml`, until we can switch from Travis to GitHub Actions / CF? I can't guarantee it will work (because the issue is due to Travis ending free builds), but I think it's worth a shot.

**Comment by roveo** [comment]
> Better check that setuptools has this functionality for all python versions of interest.
Do we have a list of python versions that should be supported? Setuptools supported entry_points since `0.6a1` (year 2009), so I don't think this should be a problem.

**Comment by roveo** [comment]
Added Plugins page to docs (I'm not a native speaker, so edits and additions are very welcome).

Note that there's a table commented out in __Known plugins__ section. I think at this point we should think about which existing functionality should go into extras (the leading candidate being kafka).

File: streamz/core.py
Comment (martindurant) [inline streamz/core.py:309]: Could use a docstring, since the real method won't appear until after it's been called at least once. 
"entrypoints-based streamz plugin, loaded on first call"

File: streamz/tests/test_plugins.py
Comment (martindurant) [inline streamz/tests/test_plugins.py:44]:
```
assert inspect.isfunction(Stream().from_test)
```
(would be a method if not static)
Note that `Stream` is instantiated.

File: streamz/tests/test_plugins.py
Comment (martindurant) [inline streamz/tests/test_plugins.py:67]: Actually, I'm not sure whether the method should have been overwritten - presumably `Stream.test_double` already existed before the call to `register_plugin_entry_point`.

In normal usage, the stubs always come first, since they happen at import time, and streamz should always be imported before derived classes.

**Review by martindurant — COMMENTED** [review]
I really like this PR

File: streamz/tests/test_plugins.py
Comment (roveo) [reply streamz/tests/test_plugins.py:44]: Nice! thanks, I spent some time figuring out how to do this better, haven't thought of `inspect`.

File: streamz/tests/test_plugins.py
Comment (roveo) [reply streamz/tests/test_plugins.py:67]: Yes, I guess this test actually doesn't test anything. I wanted to test that running `Stream.register_api` twice on the same class doesn't break anything, in case the implementation of `register_api` changes. We can just check for this in `register_plugin_entry_point`.

Comment (martindurant) [reply streamz/tests/test_plugins.py:67]: The test is OK, but not comprehensive. The cases are:
- register_api, then register entrypoint
- register entrypoint then register_api
- either of the to methods called twice

In every case, calling the method should result in the same test class.

We *could* check for whether the name is being overwritten, but I don't think that's essential.

Comment (roveo) [reply streamz/tests/test_plugins.py:67]: Btw, what should we do if a plugin wants to override a built-in method? On the one hand, this would allow people to create useful extensions of built-in functionality (my version of `partition` could be provided as a plugin long before it's merged into core), on the other — it could break things and lead to confusion.

Comment (CJ-Wright) [reply streamz/tests/test_plugins.py:67]: Personally I would have the system balk at a built-in override. My hope would be that names are cheap, so adding more text onto a would be `partition` describing how it is difficult would add less burden than issues around which `partition` was used. This could get particularly thorny as envs get updated, code that previously worked and relied on the built-in are now producing errors in subtle ways that could be difficult to find.

Comment (roveo) [reply streamz/tests/test_plugins.py:67]: > In every case, calling the method should result in the same test class.

The class is stored in `Stream.method.__wrapped__`, but I don't see how anything else but the test class can end up there in any of these cases.

Comment (martindurant) [reply streamz/tests/test_plugins.py:67]: You mean we should have a `clobber=False` argument for the register methods? 
I wonder in that case, if there is a sane way to tell when an entrypoint and register_api refer to the same thing, which would not be a problem even without clobber.

Comment (roveo) [reply streamz/tests/test_plugins.py]: > I wonder in that case, if there is a sane way to tell when an entrypoint and register_api refer to the same thing, which would not be a problem even without clobber.

We can just check that `hasattr(Stream, entry_point.name)`. But would have to happen during plugin load (so that we're not overwriting built-in methods) and so importing streamz with a bad plugin would result in an error.

Comment (martindurant) [reply streamz/tests/test_plugins.py]: but this will fail when someone adds the entrypoint to a class that already has register_api, no?

Note @CJ-Wright , that the current implementation does allow you to use register_api to overwrite methods at will. Of course this is python, so we can't ever disable someone who wants to do that anyway.

Comment (CJ-Wright) [reply streamz/tests/test_plugins.py]: That's fair. To some extent this boils down to locality for me. If the author of the code is the user of the env then if they override things then they are in a better place to clean up. When installing 3rd party things that state may not hold as well.

Comment (roveo) [reply streamz/tests/test_plugins.py]: > If the author of the code is the user of the env then if they override things then they are in a better place to clean up. When installing 3rd party things that state may not hold as well.

I agree. With `register_api` things are explicit and visible. A plugin is a black box.

@martindurant please look at the last commit to see what I mean

File: streamz/core.py
Comment (martindurant) [inline streamz/core.py:324]: I don't *think* this can be False, but it doesn't hurt to check.

**Review by martindurant — COMMENTED** [review]
I'm happy with this. @CJ-Wright ?

File: streamz/core.py
Comment (roveo) [reply streamz/core.py:324]: It is false when the plugin class returned from `entry_point.load()` is decorated with `@Stream.register_api`. Then it is registered right away when loaded, and so at that moment stub is overwritten with the actual class.

**Review by CJ-Wright — APPROVED** [review]
LGTM! One thing to consider is scraping the conda-forge packages/data for packages who declare the streamz plugin, if we are interested in building a plugin registry.

**Comment by martindurant** [comment]
Right, but there aren't any yet :)
I actually would expect any plugin developers to contact us, if they wanted to be listed on the streamz documentation somewhere.

**Comment by martindurant** [comment]
@chinmaychandak : that kafka test is failing too often! (and giving an incorrect message, due to using positional rather than keyword arguments to wait)

**Comment by CJ-Wright** [comment]
It was more of a general idea, since other plugin mediated systems could use that approach as well.

**Pull request merged**
